specs/agents/tracing-instrumentation-azure-functions.md

```diff
@@ -0,0 +1,103 @@
+
+## Generic Instrumentation
+
+In general, to instrument an Azure Functions application, we create transactions that wrap the execution of its handler methods. In cases where we cannot assume any trigger type or extract trigger-specific data (e.g. if the trigger type is unsupported),
+we wrap the handler method with a transaction, while using the generic context object to derive necessary fields:
+
+| Field | Value | Description | Source |
+| - | - | - | - |
+| `name` | e.g. `MySampleTrigger` | The transaction name. Use function name if trigger type is `other`. | *generic context* |
+| `type` | e.g. `request`, `messaging` | The transaction type. | Use `request` if trigger type is undefined. |
+| `outcome` | `success`/`failure` | Set to `failure` if a function error can be detected otherwise `success`. | |
+| `result` | `success`/`failure` |Set to `failure` if a function error can be detected, otherwise `success`. | Trigger specific. |
+| `faas.name` | e.g. `MySampleTrigger` | The function name. | *generic context* |
+| `faas.id` | e.g. `/subscriptions/d2ba53be-0815-4...` | The [fully qualified resource ID](https://learn.microsoft.com/en-us/rest/api/resources/resources/get-by-id) of the Azure Function, which has this format: `/subscriptions/<SUBSCRIPTION_GUID>/resourceGroups/<RG>/providers/Microsoft.Web/sites/<FUNCAPP>/functions/<FUNC>` | *generic context*, environment |
+| `faas.trigger.type` | `other` | The trigger type. Use `other` if the trigger type is unknown or cannot be specified. | More concrete triggers are `http`, `pubsub`, `datasource`, `timer` (see specific triggers below). |
+
+| `faas.execution` | `203621a2-62f...` | The unique invocation id of the function. | *generic context* |
+| `faas.coldstart` | `true`/`false` | A boolean value indicating whether this function invocation was a cold start or not. | [see section below](deriving-cold-starts)
+
+### Metadata
+
+Generic cloud metadata fetching SHOULD be disabled when the agent is running in an Azure Functions app. It will not
+yield any useful data and only incur additional cold start overhead due to the non-existent HTTP metadata endpoints.
+
+Agents SHOULD rather:
+
+* Check for the existence of the `FUNCTIONS_WORKER_RUNTIME` environment variable that indicates that the agent is running
+in an Azure Function application.
+* Perform specific cloud metadata detection for Azure Functions (see [Metadata](./metadata.md)).
+
+In addition to those described in [Metadata](./metadata.md), following metadata fields are relevant for Azure Functions:
+
+| Field | Value | Description | Source |
+| - | - | - | - |
+| `service.framework.name` | `Azure Functions` | Constant value for the framework name. | |
+| `service.runtime.name`| e.g. `dotnet-isolated` | The language worker runtime (see [here](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings#functions_worker_runtime)). | `FUNCTIONS_WORKER_RUNTIME` |
+
+### Deriving cold starts
```

Comment:
> Does that vary by language runtime?

https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-http-webhook-trigger?tabs=in-process%2Cfunctionsv2&pivots=programming-language-csharp#example suggests that for C# both in-process and out-of-process exists.

Reply:
> Does that vary by language runtime?

I assume it COULD vary by the language runtime and Microsoft will probably not lock itself in by promising it will be this or that way forever.
However, from what I observed so far (.NET, Node.js, Python): if a function app has multiple function entry points, these can run concurrently in the same process.

Ad ".NET 's `in-process` and `out-of-process` mode": This is not really related to that. Both of those hosting models
will allow concurrent function invocations in the same process.